sklearn/model_selection/_search.py

```diff
@@ -999,3 +1004,356 @@ def fit(self, X, y=None, labels=None):
                                           self.n_iter,
                                           random_state=self.random_state)
         return self._fit(X, y, labels, sampled_params)
+
+
+class SequentialSearchCV(BaseSearchCV):
+    """Search over hyperparameters using Gaussian process.
+
+    The parameters of the estimator used to apply these methods are optimized
+    by cross-validated search over parameter settings.
+
+    In contrast to GridSearchCV, not all parameter values are tried out, but
+    rather a fixed number of parameter settings are sampled from the specified
+    distributions by optimizing an acquisition function. The number of
+    parameter settings that are tried is given by n_iter.
+
+    If all parameters are presented as a list,
+    sampling without replacement is performed. If at least one parameter
+    is given as a distribution, sampling with replacement is used.
+    It is highly recommended to use continuous distributions for continuous
+    parameters.
+
+    Read more in the :ref:`User Guide <gp_search>`.
+
+    Parameters
+    ----------
+    estimator : estimator object.
+        A object of that type is instantiated for each grid point.
+        This is assumed to implement the scikit-learn estimator interface.
+        Either estimator needs to provide a ``score`` function,
+        or ``scoring`` must be passed.
+
+    param_distributions : dict
+        Dictionary with parameters names (string) as keys and distributions
+        or lists of parameters to try. Distributions must provide a ``rvs``
+        method for sampling (such as those from scipy.stats.distributions).
+        If a list is given, it is sampled uniformly.
+
+    n_init : int, optional
+        Number of random iterations to perform before the smart search.
+        Default is 3.
+
+    n_iter : int, default=10
+        Number of parameter settings that are sampled. n_iter trades
+        off runtime vs quality of the solution.
+
+    acquisition_function : string, optional
+        Function to maximize in order to choose the next parameter to test.
+        - 'UCB : maximize the upper confidence bound
+        - 'EI' : maximizes the expected improvement
```

Comment:
Do you think there's any interest in ultimately supporting an arbitrary `scipy.optimize.minimize` algorithm? How hard would it be to support their callback approach while keeping the code readable?